[Easy] The event_id of torch.cuda.Event and torch.xpu.Event always is 0

fffrog opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151226
* #151411
* #151221
* #151404

Although torch.cuda.Event and torch.xpu.Event have cuda_event and sycl_event fields respectively, the event_id exposed from the base class torch.Event is always 0, which can confuse users.

The memory of torch.Event is not useful to torch.cuda.Event and torch.xpu.Event, but we still need to inherit from torch.Event because CPython will check it.

Repro with cuda:
```
>>> import torch
>>> event = torch.cuda.Event()
>>> event.cuda_event
0
>>> event.event_id
0
>>> event.record()
>>> event.cuda_event
127982096
>>> event.event_id
0
```

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

FFFrog commented:
@pytorchbot label "topic: not user facing"

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

albanD reviewed (commented):
What do you mean by "always is 0" if the property doesn't exist I would expect you would get an error when trying to access it?

Also test?

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

FFFrog commented:
> What do you mean by "always is 0" if the property doesn't exist I would expect you would get an error when trying to access it?

I added a repro in the description of PR, hopfully help.

> Also test?

Added also.

fffrog committed:
Update

[ghstack-poisoned]

albanD reviewed (commented):
Why isn't the fix to update the c10::Event to have the appropriate eventId() impl for these devices?

fffrog committed:
Update

[ghstack-poisoned]

FFFrog commented:
> Why isn't the fix to update the c10::Event to have the appropriate eventId() impl for these devices?

This pr method should be the simplest and most effective way to solve this problem, but it's a bit ugly. :D

Possible methods:
- Refer to the `torch.Stream` method, and modify the corresponding fields in `torch.Event` at the same time where `torch.cuda.Event` or `torch.xpu.Event` creates `event_` (the disadvantage is that the modification points are scattered and there are many entries)
- If we don't modify `torch.cuda.Event `or `torch.xpu.Event`, but only modify the `event_id` implementation of `torch.Event`, we need to distinguish the memory layout of `_self` in the `THPEvent_evend_id` to access the corresponding correct fields (the modification is also relatively large and the scalability is not very good)

I don`t know whether there are some other better way to fix it, any suggestions?

albanD reviewed (approved):
Ho I though this was also for the torch.Event, but it's just for these one offs.
Sounds good to have a weird thing here then, we recommend users don't use these anymore anyways.

FFFrog commented:
> Ho I though this was also for the torch.Event, but it's just for these one offs.
> 
> Sounds good to have a weird thing here then, we recommend users don't use these anymore anyways.

I knew, Thank you a lot

FFFrog commented:
@pytorchbot merge

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

FFFrog commented:
@pytorchbot merge -i

FFFrog commented:
@pytorchbot merge -i

FFFrog commented:
@pytorchbot merge -i

malfet commented:
@pytorchbot revert -m "Reverting to unblock revert of https://github.com/pytorch/pytorch/pull/151404" -c weird

fffrog committed:
Update

[ghstack-poisoned]

fffrog committed:
Update

[ghstack-poisoned]

FFFrog commented:
@pytorchbot merge -i
